third_party/move/move-vm/runtime/src/loader/function.rs

```diff
@@ -140,20 +148,84 @@ impl LazyLoadedFunction {
+        traversal_context: &mut TraversalContext,
         fun: Rc<LoadedFunction>,
         mask: ClosureMask,
     ) -> PartialVMResult<Self> {
+        let runtime_environment = module_storage.runtime_environment();
         let ty_args = fun
             .ty_args
             .iter()
             .map(|t| runtime_environment.ty_to_ty_tag(t))
             .collect::<PartialVMResult<Vec<_>>>()?;
+
+        // When building a closure, if it captures arguments, and it is persistent (i.e., it may
+        // be stored to storage), pre-compute layouts which will be stored alongside the captured
+        // arguments. This way layouts always exist for storable closures and there is no need to
+        // construct them at serialization time. This makes loading and metering logic much simpler
+        // while adding layout construction overhead only for storable closures.
+        let captured_layouts = fun
+            .function
+            .is_persistent()
+            .then(|| {
+                Self::construct_captured_layouts(
+                    module_storage,
+                    gas_meter,
+                    traversal_context,
+                    &fun,
+                    mask,
+                )
+            })
+            .transpose()?;
+
         Ok(Self(Rc::new(RefCell::new(
-            LazyLoadedFunctionState::Resolved { fun, ty_args, mask },
+            LazyLoadedFunctionState::Resolved {
+                fun,
+                ty_args,
+                mask,
+                captured_layouts,
+            },
         ))))
     }
 
+    /// For a given function and a mask, constructs a vector of layouts for the captured arguments.
+    pub(crate) fn construct_captured_layouts(
+        module_storage: &impl ModuleStorage,
+        _gas_meter: &mut impl DependencyGasMeter,
+        _traversal_context: &mut TraversalContext,
+        fun: &LoadedFunction,
+        mask: ClosureMask,
+    ) -> PartialVMResult<Vec<MoveTypeLayout>> {
+        let ty_converter = StorageLayoutConverter::new(module_storage);
+        let ty_builder = &module_storage.runtime_environment().vm_config().ty_builder;
+
+        mask.extract(fun.param_tys(), true)
+            .into_iter()
+            .map(|ty| {
+                let (layout, contains_delayed_fields) = if fun.ty_args.is_empty() {
+                    ty_converter.type_to_type_layout_with_identifier_mappings(ty)?
+                } else {
+                    let ty = ty_builder.create_ty_with_subst(ty, &fun.ty_args)?;
+                    ty_converter.type_to_type_layout_with_identifier_mappings(&ty)?
+                };
+
+                // Do not allow delayed fields to be serialized.
+                if contains_delayed_fields {
+                    let err = PartialVMError::new(StatusCode::VALUE_SERIALIZATION_ERROR)
+                        .with_message(
+                            "Function values that capture delayed fields cannot be serialized"
```

Comment:
Not a good error message if we actually haven't serialized anything. Perhaps just set capture layouts to None in this case.